src/numericalAnalysis.cpp

```diff
@@ -0,0 +1,119 @@
+#include "mantella_bits/numericalAnalysis.hpp"
+
+// C++ standard library
+#include <cmath>
+#include <limits>
+
+// Mantella
+#include "mantella_bits/assert.hpp"
+
+namespace mant {
+  double brent(
+      std::function<double(double)> objectiveFunction,
+      double lowerBound,
+      double upperBound,
+      double xTolerance) {
+    double lowerBoundObjectiveValue = objectiveFunction(lowerBound);
+    double upperBoundObjectiveValue = objectiveFunction(upperBound);
+
+    verify(lowerBoundObjectiveValue * upperBoundObjectiveValue <= 0.0, "brent: The lower bound and upper bound objective value must have a different sign.");
+
+    double previousUpperBound = lowerBound; // only on initialisation: previousUpperBound = lowerBound
+    double previousUpperBoundObjectiveValue = lowerBoundObjectiveValue;
+    double boundDistance = upperBound - lowerBound;
+    double toleranceCriteria = boundDistance;
+
+    double middleOfBounds;
+    double boundObjectiveValueRatio;
+    double interpolationNumerator;
+    double interpolationFactor;
+    double boundPreviousBoundRatio;
+
+    int iter = 0;
+    int maxIter = 1000;
+    double relativeTolerance;
+
+    while (iter++ < maxIter) {
+      if (upperBoundObjectiveValue * previousUpperBoundObjectiveValue > 0) {
+        previousUpperBound = lowerBound;
+        previousUpperBoundObjectiveValue = lowerBoundObjectiveValue;
+        boundDistance = upperBound - lowerBound;
+        toleranceCriteria = boundDistance;
+      }
+
+      if (std::abs(previousUpperBoundObjectiveValue) < std::abs(upperBoundObjectiveValue)) {
+        lowerBound = upperBound;
+        upperBound = previousUpperBound;
+        previousUpperBound = lowerBound;
+
+        lowerBoundObjectiveValue = upperBoundObjectiveValue;
+        upperBoundObjectiveValue = previousUpperBoundObjectiveValue;
+        previousUpperBoundObjectiveValue = lowerBoundObjectiveValue;
+      }
+
+      relativeTolerance = 2.0 * std::numeric_limits<double>::epsilon() * std::abs(upperBound) + xTolerance;
+      middleOfBounds = (previousUpperBound - upperBound) / 2.0;
+
+      if (std::abs(middleOfBounds) > relativeTolerance && upperBoundObjectiveValue != 0.0) {
```

Comment:
This inverted break condition should be commented.

Checking `upperBoundObjectiveValue != 0.0` does not seem to be numerically save. I would advise to replace this by `std::abs(upperBoundObjectiveValue) > 1e-12`.